Add native asset caching to Service Worker

Jaifroid opened the pull request:
This addresses ~~#441~~ #414. It adds the native Cache API to `service-worker.js` for the caching of css and js assets. It significantly speeds up SW mode. The cached entries can be checked in the Cache entry of devtools in Chromium or Edge (I didin't try with Firefox). Example of running this on the most recent Ray Charles is shown in the image below (from Chromium / Edge 78). This is not for merging yet, but I would appreciate testing and any comments.

![image](https://user-images.githubusercontent.com/4304337/63449072-37cdf280-c437-11e9-8453-422ee6115da6.png)

I think the main TODO is to provide a mechanism for deleting / pruning entries. We could just do it on change of ZIM, but as the quota on most devices is quite generous, and as the entries are unique to each ZIM, a more subtle mechanism could be found. We could provide usage statistics on the Configuration page and a button to clear the cache if it is quite large, for example. This would need to be decided.

Info for testing: To delete the cache and Service Worker after testing (in Chromium), delete `kiwixjs-cache` (right-click and delete) and unregister the Service Worker (under Application -> Service Workers ... ).

If there are code updates, a similar procedure would need to be followed, then a full page refresh (e.g. Ctrl-Shift-R), before testing again.

mossroy commented:
I suppose this addresses #414, not #441 (wrong link in the description).
As you know, I believe the use of emscripten version of kiwix-lib/libzim is the right way to improve performance (and libzim already has its own cache).
On the other hand, implementing #414 might be easy, maybe easier than #411 or #415. And it could still be useful with kiwix-lib/libzim.
So why not, if it's easy to unplug later (it seems to be the case), if we feel it becomes useless in the future.

It's hard to see the changes in github, that tends to say that everything has been modified.
But I saw the most important parts, which look good.

I quickly tested on Firefox, and it seems to work well too (I see the cache content in devtools, and the log properly says it uses it). This is promising.
I tested in a Chromium extension, but it gives some errors :

> Uncaught (in promise) TypeError: Request scheme 'chrome-extension' is unsupported
>    at service-worker.js:192

It fails because we can not `cache.put` a request with this scheme.
I don't know if it would be the same in Firefox (with a schema 'moz-extension') because, for now, the SW mode is not available at all with Firefox extensions.
I don't know what is the best way to handle that : blacklist this scheme in the SW cache code? But maybe Google might allow it in the future, who knows.
If we put a whitelist instead, for specific schemes (like `file`, `http` and `https`), we might miss the opportunity to activate the cache if Mozilla allows it (or in some other context).
Maybe we could trap the error, keep inside the SW a list of schemes that failed, add an info log, and then do not attempt to cache any of these schemes. But maybe it becomes a bit complicated.

The Cache API is not available on all platforms, see https://developer.mozilla.org/en-US/docs/Web/API/Cache#Browser_compatibility. I think we don't have any supported platform that supports SW, but not Cache API. But, to be cleaner, it would be better to test for this API availability before using it. If it's not available, we can simply fallback to standard behavior (without cache), and put an information message in the console.

Regarding the cache eviction strategy, things are quite open.
I agree that, in any case, we should add a button in settings to manually clear the SW cache. If it's easy to add the statistics, it would be great.
I also agree that we can keep cache from different ZIM files (not necessary to clear the cache on ZIM change)
If we use the cache only for javascript and CSS, maybe it's not necessary to add an eviction strategy, and simply rely on the standard strategy of the browser (it can clear the cache if it feels it's necessary, and can ask the user if he allows to use more cache in case it becomes big). Because these files are not numerous and won't use a lot of space, it might be enough. Except for PhET, that we should test more.

But we also might activate the cache for some small images (that could be useful for some icons), for which a specific eviction strategy could be used (with a smaller available space). But well, that could be for later, and maybe too complicated for the need.

Jaifroid commented:
I'm pleased you're interested in this, @mossroy ! It really does make a relatively easy performance jump in SW mode, while we wait for #116. Up till now, jQuery mode has been much faster than SW mode, due to the fact that we have memory caching for jQuery, but none for SW (and also because SW decompresses scripts). This equalizes things, and makes using SW mode much more feasible for daily use.

It's a real shame that the Chrome Extension schema doesn't work with Cache, as Chrome Extension is a major use-case for this app. I wonder if there's a solution for that, or whether it's just outright not supported. It's worth investigating.

Regarding an interface to control the Cache, I did develop one when I worked on a comprehensive cache based on indexDB, and it looks like the image below. This could be adapted to our needs.

I didn't know libzim comes with a cache. But it'll take some time for that code to be developed, so I think it's worth using this optimization.

I agree PhET is a problem. JS is unique to each experiment and shouldn't be cached. I'm sure we can find a way to distinguish.

Thanks for the detailed and thoughtful comments, which all seem very sensible.

![image](https://user-images.githubusercontent.com/4304337/45638336-d245e900-baa4-11e8-8f8e-514f5f52083d.png)

Jaifroid commented:
I've checked support in chrome-extension, and unfortunately it's not supported full stop. Chrome-extension has its own caching strategy that devs can use via `chrome.storage`, but this is only available in the background script, which seems like a huge complication. People seem to recommend using localStorage or indexedDB as an alternative.

Because localStorage is synchronous, it cannot be accessed from a Service Worker. IndexedDB is, however, asynchronous and apparently can be used inside Service Worker.

**So the question is: is it worth adding a fallback of indexedDB caching in SW specifically for Chrome extension usage?** I have a ready-made indexedDB function.

It should be noted that the Cache API does *not* need to be accessed from the SW. It is also available in the window object, and can be accessed from the main thread. However, it seemed to me to be neater to keep it all in the SW (also running on a separate process, so maybe faster).

Jaifroid commented:
For now, I've added a (volatile) memory cache (`assetsCache`) to the code for excluded URL schemata. This works well to cache assets during a session, just like cssCache in the main thread, which should be a gain for Chrome extensions.

To test, see the comment in the new code (commit https://github.com/kiwix/kiwix-js/pull/556/commits/5f37306e1690609b02cc169a22dcc09899a0588f) about changing `example-extension` to `http` and running the code on localhost. Remember to clear any existing Service Worker first, as it won't auto-update. You should then get messages in console log showing "[SW] Adding EXCLUDED schema URL...". Shown below is the snapshot of `assetsCache` (which is just a Map). Note we can't actually store a Response in a Map because Responses are special objects that get annulled easily (I wasn't successful storing and retrieving them). Instead I store the constituents of the Response we construct.

![image](https://user-images.githubusercontent.com/4304337/63613962-c9756580-c5d9-11e9-9fc5-d6cc5bd86979.png)

Jaifroid commented:
Ah, the dreaded Firefox 45 Travis error...

mossroy commented:
No, it's not necessary to add anything when inside the chrome-extension scheme, because Chromium already has a built-in in-memory cache (see https://github.com/kiwix/kiwix-js/issues/411#issuecomment-511199358) when inside an extension.
It should even not go through the SW more than once, and bypass it to read its in-memory cache.
No need to make this more complicated.

I think we simply need to properly avoid the console error messages in this case

Jaifroid commented:
Ah OK, I'd forgotten about that comment (haven't used Chrome extension much). So should I remove the memory chache I've added in the last commit? (It only took me 30mins or so to add it, so no great loss if you don't see a use for it.) The only use I can think of might be to speed up cache access generally.

Jaifroid commented:
What I mean by that is that, when I was experimenting with IndexedDB last year, I found it was noticeably slower to access assets stored there than from cssCache. So I opted to keep cssCache for fast retrieval, and use IndexedDB to store assets between sessions. But I'm not sure it's the case with Cache API. It seems "fast enough".

mossroy commented:
IMHO you should remove this last commit and only keep the Cache API (without introducing IndexedDB either)

Jaifroid commented:
> IMHO you should remove this last commit and only keep the Cache API (without introducing IndexedDB either)

Done. I've kept the structural improvements to the code: providing the relevant regexes at the top of the code with a comment for developers on what they do. It's better than having regexes buried further down the code which would be difficult to maintain.

NB this code excludes `^chrome-extension:` from caching, so we don't get the console log message (as you suggested above) and unnecessary CPU cycles. It can easily be changed in the future if extensions start to support Cache. See comments at top of file.

Jaifroid commented:
Just to say that it would be best for me to delay working on adding a cache control panel (like that above) until we've merged #527, because otherwise I'd have to do double work (Bootstrap 3 panels are rather different from Bootstrap 4 cards). But in due course it would be good to know whether that design seems right, and whether you'd be interested in having the "Return to last visited page when you open an archive" functionality. That functionality is useful, for example, in a NWJS app. It's less useful if the user still has to pick the archive, which is the case in most contexts. The Cache used field would show "Memory cache" for jQuery mode and "Cache API" for SW mode.

Jaifroid commented:
I had some time yesterday evening, so added the cache evacuation infrastructure and UI anyway. The checkbox for remembering the last page between sessions is currently *non-functional* and can be removed if that functionality is not needed / wanted, in which case the panel title would change from "Performance and privacy settings" to "Performance settings". Asset counting works for both the memory cache and Cache API, but only the assets in the respective cache being used are shown. However, on deleting assets, the sum of all assets deleted is displayed.

Jaifroid commented:
As a side issue, the "random" Firefox 45 Travis failures must I guess be caused by a race condition somewhere in our code. Is there a way to debug Travis builds? I suppose the Chrome failure for #527 is also a race condition (but a different one).

mossroy commented:
It would be better to discuss remembering the last page in a different PR

Jaifroid commented:
> It would be better to discuss remembering the last page in a different PR

OK, I've added issue #561 for this.

Jaifroid commented:
Remember last page checkbox removed. This is what the interface looks like straight after the user turns caching off:

![image](https://user-images.githubusercontent.com/4304337/63914238-7876ce80-ca2a-11e9-9459-afa22cc72242.png)

Jaifroid commented:
I've rebased this PR on master and have adapted the cache control panel to use Bootstrap 4.

Jaifroid commented:
I think this PR is ready for review, unless there are any further features that need to be added. The only thing that occurs to me is that, in accordance with #507, I could put the new functions in `app.js` into a new module, called `cache.js`. Would that be worthwhile?

mossroy commented:
I quickly tested on Firefox and Chromium and like the settings UI.
If it's technically possible, it would be better to add how much space the cache takes (and the quota of the browser). For the Cache API, https://developer.mozilla.org/en-US/docs/Web/API/StorageEstimate should do it. For the Memory cache, it's less necessary as it is emptied at the end of each session.
I did not have a deep look at the code, but I'm puzzled by the fact that you are able to display the "Caching _filename_ ..." overlay in SW mode. It's consistent with the jQuery mode, which is cool, but it probably means you're modifying the DOM when getting some content for the SW, which is not cool.
The code that is ran at this moment can not know if the asset will be put in cache or not, which has a side-effect when running as a chromium extension : on every page it says it is caching the CSS and javascript files (which is wrong).
In this extension context, the settings UI says it uses a "Memory" cache, but it always has zero assets, which can be confusing. The truth is that kiwix-js does not use any of its own cache (memory or CacheAPI) BUT Chromium caches all the files by itself (in another memory region, and it can probably not be detected in javascript). Maybe it's not crucial if we don't give a very precise information here, but we must try to not give a wrong one.

IMHO, we should not display "Caching ..." in SW mode because it breaks the software layers, and can not be reliable. Never mind if it's not consistent with jQuery mode in this case.
And the Settings UI should be fixed for the case of Chromium extension.
The difficulty is that we would have to duplicate the logic that is in the SW (blacklisting chromium extensions). I'm wondering if a better approach would be to test the SW instead of duplicating some of its code : asking for a dummy content, and see if the SW serves it from cache or not (this information could be added in the response headers by the SW). But maybe it's too complicated for the current need.

mossroy commented:
Thinking again about that, I changed my mind.
I think almost nobody will care about this setting, and people who care probably can use the devtools to have more info.
I would prefer that we keep this implementation as simple as possible.

So I suggest that : 
- we don't try to get the space usage of caches (only keep the count you already have)
- we change a bit the behavior of the UI so that we don't need to tell the user which cache is in use : we could simply always display both caches count in the UI, and empty both of them when the user clicks to disable them
- we don't try to guess which one will be in use

Jaifroid commented:
> I'm puzzled by the fact that you are able to display the "Caching filename ..." overlay in SW mode. It's consistent with the jQuery mode, which is cool, but it probably means you're modifying the DOM when getting some content for the SW, which is not cool

He he, nothing gets past you! ;-) It's very easy: it's just [this one line](https://github.com/kiwix/kiwix-js/pull/556/files#diff-bca463ed8ccf5e2058dfe307a95e8501R1065) linking to the same function that updates the cache status messages in jQuery mode. I can remove the line if you really think it's problematic -- personally I find it helpful to have feedback about why a load is taking so long. It helps to see that the app hasn't got stuck (which it still does occasionally, especially when switching between jQuery and SW modes).

Jaifroid commented:
PS The code doesn't need to know if the cache is being used or not: if it is being used, the file won't be extracted by readBinaryFile and the message won't be shown. If the file is being extracted by `readBinaryFile`, then no cache (not even Chrome extension's built-in cache) has been used for the file. It would be easy to exempt showing the message if we are in a Chrome extension. But I'd like to investigate why its in-built caching is clearly not working in that case, as we don't want unnecessary extractions from the ZIM.

Jaifroid commented:
PPS Maybe some confusion is caused by the fact that the message says "Caching ...". More accurately, it should say "Extracting ..." or "Getting ..." (regardless of whether we keep it for SW mode or not). The message is not registering a cache, it's registering extraction of the file from the ZIM archive, but only uncached files are extracted.

Jaifroid commented:
So, with the latest updates (which I've tested in extension form on Edge Chromium) we have the following simplified behaviour:

* In both jQuery and SW modes, the Cache count shown is always the sum of cssCache.size and CacheAPI count;
* If the total is zero and we're in a Chrome extension, the asset count is shown as '-', so as not to mislead the user, but if the caches we control do have something in them, that amount is shown;
* No caching overlay is shown in SW mode in a Chrome extension;
* Caching overlay is currently shown in other contexts in both jQuery and SW modes, but removing it from SW mode is just a case of removing one line.

The logic for working out which cache we're using is already written, and it's not complicated or a maintenance headache IMHO. If chrome-extension starts using the standard Cache, it's a simple edit to standardize it, and I've added comments in the appropriate places in SW mode and jQuery mode (with cross-referencing) for future Devs (most likely ourselves) to remember what to do.

So hopefully that addresses your points. It remains to be decided if I should change "Caching ..." to "Extracting ..." or "Getting ..." (latter a bit shorter), and for you to confirm that you really, really don't want the pretty overlay in SW mode 😉.

mossroy commented:
To me, we have unneeded logic and assumptions, where we could simply have a UI like this (instead of the "Cache status" box) :

```
Cache usage (# of assets) : 
Memory cache : 18      Cache API : 23
```
Without telling the user which cache is currently in use. It avoids making some assumptions (that might become wrong in the future) and some duplicated code/logic with the SW.

On the left of this box, when the user clicks on "Do not cache assets", it empties both caches, and refreshes the cache usage box, so that both show zero assets (no need to tell how much have been removed)

When the user switches between jQuery and SW mode, nothing special happens regarding this cache status

It seems much simpler to me, and much easier to maintain. The drawback is that the user does not know which cache region is currently in use. But we don't know for sure either (we did not test every context where it can be run). And there is always a workaround to know for sure : open a page, come back there and see which figures have increased.

Regarding the wording, "Getting ... " is fine for me, but you're the expert on that topic ;-)
And yes, I still think we must not modify the DOM while getting some content for the SW.

mossroy commented on service-worker.js:
```diff
@@ -23,117 +23,191 @@
 'use strict';
 
-self.addEventListener('install', function(event) {
+var CACHE = 'kiwixjs-cache';
```
This should be renamed to indicate that it's a cache of assets, used by the SW

Jaifroid replied:
`kiwixjs-assetCache`? NB The Cache API could perfectly well be used in jQuery mode. The fact that we only implement it in SW mode is an arbitrary decision (albeit a sensible one for us at this moment in time). Cache API is available in `window`, so accessible from app.js.

mossroy replied:
All right. This name looks fine

Jaifroid replied:
I've changed the name. Be sure to delete the old Cache, either before downloading the commit, or in dev tools.

mossroy commented on service-worker.js:
```diff
@@ -23,117 +23,191 @@
+                        // Add css or js assets to CACHE (or update their cache entries) unless the URL schema is not supported
+                        if (cachedContentTypesRegexp.test(response.headers.get('Content-Type')) &&
+                            !excludedURLSchema.test(event.request.url)) {
+                            event.waitUntil(updateCache(event.request, response.clone()));
```
Why is it necessary to wait for the cache to be updated before returning the response? (2 lines below).
It should not be a problem that the cache can be updated a bit later, and should speed up the request/response timing?

Jaifroid replied:
@mossroy It doesn't wait. `event.waitUntil` may look like async `await`, but in fact it is merely an instruction to the browser not to let the event expire until the cache has been updated (events are expired quickly by browsers to optimize memory usage). The next line (`return response;`) will run immediately, before the `updateCache` promise completes, and we can't `return` before the `event.waitUntil` statement or else that code will become unreachable.

mossroy replied:
All right, thanks for the explanation

Jaifroid commented:
> To me, we have unneeded logic and assumptions, where we could simply have a UI like this (instead of the "Cache status" box) :
> Cache usage (# of assets) : 
> Memory cache : 18      Cache API : 23

This would be possible. But the reason that I originally implemented it in a more dynamic way was because I wanted a concise display that could adapt easily to whatever cache had been implemented or was supported by the current browser: Memory / LocalStorage / IndexedDB / Cache API. My original implementation had Memory, LocalStorage and IndexedDB. I'd quite like to keep compatibility of the codebase with future versions of Kiwix JS Windows (2.0) where I'd like to adopt this interface (I haven't ported it there yet) but where Service Worker is not available and Cache API is available only on some devices, so I will most likely use IndexedDB, (and Cache API in environments that support it).

If you feel strongly the interface needs to be simplified, it wouldn't be hard to change it to your preferred model above, but it would complicate the relationship between the two code bases as I'm envisioning it.

> And yes, I still think we must not modify the DOM while getting some content for the SW.

Personally I would distinguish *conceptually* between "modifying the DOM" and "modifying the UI", even though they are *technically* the same thing in HTML5 development. The former sounds scary, but all we're doing is providing UI feedback. However, I respect your strong views on this and shall remove the offending line ;-)

Jaifroid commented:
> no need to tell how much have been removed

I've deleted the functions / div that display the number of deleted assets. Instead, the asset count (which is always now a total of assets in cssCache and CACHE) is simply zeroed.

> Regarding the wording, "Getting ... " is fine for me

Since we're no longer showing this for SW mode, maybe we don't need to change the wording. In jQuery mode, I think we do actually only show this when caching, or at least very close to the `Map.set()` method, so I think it's not misleading. If we do change it, however, I think I would prefer "Fetching ..." because "Getting ..." sounds a bit like we dumbed things down (!).

mossroy commented:
Choose the wording you prefer, no problem.

About the cache status, my goal is mainly to avoid making the code more complicated, duplicated and take the risk to give a wrong information to the user.
Besides that, no "normal user" cares about how the cache behaves and is implemented. So it's no use duplicating code to give this kind of information.

Having an option to disable the cache can still be a good idea for debugging purpose (or in some edge cases), but it's still an "expert setting", and could be moved in the "expert settings" zone.

I understand your concern about code diverging between kiwix-js and kiwix-js-windows.
What about, in the case of kiwix-js-windows, using something like this on the right : 
```
Cache usage (# of assets) : 
Memory cache : 18
Cache API : 23
IndexedDB : 12
LocalStorage : 26
```
When some API is unavailable, or unusable, you could simply put a "N/A" instead of the asset count
But maybe I'm missing something, as I don't really know what it is like (I did not see it on https://kiwix.github.io/kiwix-js-windows)

Regarding PhET, it's a good use case to test what happens when many assets are put in the cache (at some point, the browser might ask if the user wants to grow the cache or not). But we must not do anything that is PhET specific.
I don't currently see a generic and reliable way to distinguish the javascript files that should be kept for long (like the ones of wikimedia) and the ones that should not (like the ones of PhET).
Except by implementing our own LRU eviction strategy, or something similar.
But I think it's not necessary to do that for now. Let's see how it behaves.

Based on https://developers.google.com/web/fundamentals/instant-and-offline/web-storage/offline-for-pwa , the browsers have their own safe-guards when too much cache is used : it might be enough to rely on them.

Jaifroid commented:
OK, I'll have a think about what can be done to simplify. As it is, we currently have three different caching strategies: "Memory", "Cache API", and "Custom" (Chrome-Extension, which is a black box). Unless we go for some automatic cache evacuation strategy, I'd like users to be aware that items are being cached potentially for a long time in the case of Cache API and that they can end up taking significant space.

Jaifroid commented:
PS This system is not implemented in Kiwix JS Windows yet. My plan is to replace the file-based caching currently used there (hard to maintain) with this automatic one (much easier to maintain), when it's ready. But I also have to provide indexedDB for jQuery mode because of the lack of general support for Service Worker in UWP JS apps (SW support only exists for content served from https: , i.e. if I implement it as a PWA inside UWP).

Jaifroid commented:
So this is my best attempt to prevent duplication of variables, regexes, and functions across app.js and SW, and to simplify the logic. The latest code works like this:

* When app.js needs to display cache details, it queries the Service Worker, if it is available, to determine the attributes of the cache in use (type, description, assetCount);
* If we are in jQuery mode, then it merely returns the attributes of cssCache (because it cannot presume that it can access SW, even if sometimes it technically could);
* When user turns off caching, all caches are deleted (except the internal Chrome extension cache of course);
* When user switches from jQuery mode to SW mode, cssCache is emptied, as it is no longer needed;
* When user switches from SW mode to jQuery mode, CACHE is emptied, as it will no longer be used, and we should not have items taking significant space that are useless (and that the user may not be aware of) [I realize this is an assumption, but I think it's correct from a user perspective, as users will not switch modes frequently, and the app should not hog memory or storage space];
* **Therefore**, the cache count only shows the count of the cache in use, or '-' if it cannot be detemined (i.e. Chrome extension). It would be useless to show any other counts, because they will always be **0**;
* CACHE is defined in app.js, and is passed to Service Worker early in its initialization. It has to be this way round, because otherwise we have no way to delete CACHE from app.js when Service Worker has been deactivated (we *need* the CACHE constant). **But we only define CACHE once.**

I think this simplifies the decision logic significantly. However, the amount of code in app.js has not decreased by much, because messaging the Service Worker and receiving a response is not completely trivial. **The big benefit is that we no longer have to tell DEVs to ensure that duplicated CACHE definitions and regexes have to be kept in sync, because there is no duplication.** So the main benefit is maintainability.

Please let me know what you think @mossroy.

mossroy commented on service-worker.js:
```diff
@@ -23,117 +23,240 @@
+function fromCache(request) {
+    // Prevents use of Cache API if user has disabled it
+    if (!useCache) return Promise.reject('disabled');
+    return caches.open(CACHE).then(function (cache) {
```
You should test that the Cache API is available before using it, so that it does not break if it's not.
It's probably a rare case, anyway

Jaifroid replied:
OK, will do.

Jaifroid replied:
Actually, thinking about this, a test might be redundant as it seems Cache API is a *required part* of the Service Worker spec. According to the [CacheStorage part of the Service Worker spec](https://w3c.github.io/ServiceWorker/#cachestorage-interface), "The user agent _must_ create a CacheStorage object when a Window object or a WorkerGlobalScope object is created and associate it with that global object."  So if we are in Service Worker the Cache API is available, even if we can't actually save a request in it due to having the wrong URL schema. Do you agree @mossroy?

mossroy replied:
You seem to be right. In this case, it's only necessary to test when in app.js

mossroy commented on www/js/app.js:
```diff
@@ -311,7 +346,62 @@ define(['jquery', 'zimArchiveLoader', 'util', 'uiUtil', 'cookies','abstractFiles
+    /**
+     * Queries Service Worker if possible to determine CACHE capability and returns an object with cache attributes
+     * If Service Worker is not available, the attributes of the memory cache are returned instead
+     * @returns {Promise} A Promise for an object with cache attributes 'type', 'description', and 'count'
```
The `type` does not seem be currently used. I suppose it's for some future usage?
If you have something in mind, it's fine : I just want to check it's not a mistake

Jaifroid replied:
Ah, my editor (vscode) shows me details for the Promise interface as if it were a valid type. But I think it's heavily orientated towards TypeScript, where strong typing is mandatory. I'll just change this to `{Object}` and make clear that it's a promise for an object in the description. We do have some custom non-JS types in use -- e.g. we have `@type Integer` elsewhere, whereas strictly there is no such primitive JS type, and we should use `{Number}`.

mossroy replied:
I was not mentioning this annotation (which seems right), but the attribute`type` of the returned object (next to `description` and `count`)

Jaifroid replied:
Ah, OK. I'm not using "type" as a technical term. It's the name of a key, so if I assign this returned object to a variable like `cache` I can use `cache.type`, `cache.description` and `cache.count`. The string `cache.type` could be reported as the strings `cacheAPI`, or `memory`, or `custom` or even `none`, depending on context and user options. Are you OK with `type` as the name of this object key? Type and description are subtly different. Type would be `cacheAPI` and description is `Cache API` with a space and initial capital (a future DEV might want a fuller description, for example).

Jaifroid replied:
Yes, you're right I don't now use it. I did before, and that code has been cut in the simplification. Should I keep?

mossroy replied:
I'm OK with the `type` name, and understand the purpose. But I did not see where this attribute was used.
It's not a big issue to keep it if you think it might be useful in the future

Jaifroid replied:
OK, I'll keep it. It should be useful for extensibility and I'm averse to having data structures that use descriptions as primary keys (from working with relational databases)!

Jaifroid commented on service-worker.js:
```diff
@@ -23,117 +23,240 @@
+ * The value is defined in app.js and will be passed to Service Worker on initialization (to avoid duplication)
+ * @type {String}
+ */
+var CACHE;
```
OK, will do.

Jaifroid commented on service-worker.js:
```diff
@@ -23,117 +23,240 @@
+            fetchCaptureEnabled = false;
+        }
+    }
+    if (event.data.useCache) {
```
I did consider that, but the problem was (if I remember correctly) that event.data.action is a string (or at least it is tested as a string) and I wanted a psuedo-Boolean. However, I can think of a solution that uses event.data.action without having to have two code blocks, so will try that, and if it's easy, will do it in this PR.

mossroy replied:
It's really not crucial, and my initial implementation might not be the best

Jaifroid replied:
> It's really not crucial, and my initial implementation might not be the best

It turns out to be very easy to do (I'm about to push a commit with it) and it does make sense to have all "actions" under this key.

Jaifroid commented on service-worker.js:
```diff
@@ -23,117 +23,240 @@
+ * Stores or updates in CACHE the given Request/Response pair
+ * @param {Request} request The original Request object
+ * @param {Response} response The Response received from the server/ZIM
+ * @returns {Promise} A Promise for the update action
```
Here we have another {Promise} return type, but it's trickier to know what to substitute, because it's a promise for an action. Looking at the https://developer.mozilla.org/en-US/docs/Web/API/Cache/put , they say that `cache.put()` returns a Promise. Is it OK to leave this here, @mossroy ? Where an obvious Object or Array is returned *via* a Promise, I'm changing the return types, but here it doesn't seem possible.

mossroy replied:
I agree

Jaifroid commented:
Commit https://github.com/kiwix/kiwix-js/pull/556/commits/dd18a67e5e75a33d8acc0d947f408a9bd1e6d34e implements what was discussed in code review. Given that some changes arose from a small misunderstanding over return types, what's your view on whether a Promise should always be described as a `{Promise}` in a JSDoc return type annotation, or whether it is better, if we are returning an Object via a Promise to describe the return type as `{Object}` and then make clear in the description that it's via a Promise. I read people saying we should only use "Primitive JavaScript types", but that may be outdated, and I can't find JSDoc official advice.

Jaifroid commented:
Last question: should I remove console.log lines (not console.error) from the code before merging, or is it useful to have them for a while and remove them later?

Jaifroid commented:
It seems the syntax `{Promise<T>}` has become the accepted usage. (And VSCode JSLint recognizes this as valid.) See https://github.com/kiwix/kiwix-js/pull/556/commits/301bee8e72e763c6dbe20e8520ec56aceb9558b1 .

mossroy commented:
Maybe we could keep the `console.debug` lines for now, and remove them later.
Regarding the way to write annotations for Promises, I don't know. Our existing code usually uses `@return {Promise}` followed by a textual description of what it will return.
If there is a more common and standardized way of doing it, please do (and maybe open another issue to refactor the existing code)

mossroy commented on service-worker.js:
```diff
@@ -208,26 +208,27 @@ function removeUrlParameters(url) {
+ * Looks up a Request in CACHE_NAME and returns a Promise for the matched Response
+ * @param {Request} request The Request to fulfill from CACHE_NAME
+ * @returns {Promise<Response>} A Promise for the cached Response
+ * @throws {Promise<String>} A rejected Promise with strings 'disabled' or 'no-match'
```
I don't think we should use the `@throws` annotation for the case of a rejected Promise.
I suppose it is meant only for exceptions that are synchronously sent by the function (like in Java) : https://jsdoc.app/tags-throws.html

In this case, we might add this information (about what is rejected by the promise) as a comment at the end of the `@returns` line

Jaifroid replied:
OK! Will remove before merging. It's somewhere in that long thread mentioned in #568, but you're right that it might not be the best way to do it, and it's not properly documented anywhere.

Jaifroid commented:
After a bit more testing (including in Chrome extension, but not Firefox extension because I don't know how), can I squash and merge?

mossroy commented:
Yes you can
